New design based on `*mut T`

phil-opp opened the pull request:
TODO:

- [x] update tests
- [ ] update docs 

cc https://github.com/rust-osdev/volatile/pull/13#issuecomment-842455552

phil-opp committed:
New design based on `UnsafeCell`

bjorn3 commented on src/lib.rs:
````diff
@@ -78,95 +80,38 @@ impl<R> Volatile<R> {
     /// assert_eq!(volatile.read(), 1);
     /// ```
-    pub const fn new(reference: R) -> Volatile<R> {
+    pub unsafe fn new<A>(reference: &UnsafeCell<T>, access: A) -> Volatile<'_, T, A> {
````
I think this is safe.

phil-opp replied:
Thanks for taking a look! This function could be safe, but I think we would need to make all read/write functions unsafe then because e.g. race conditions could occur if the `UnsafeCell` is also used in other threads.

bjorn3 commented on src/lib.rs:
```diff
@@ -78,95 +80,38 @@ impl<R> Volatile<R> {
-            reference,
-            access: PhantomData,
-        }
+    pub fn from_ref(reference: &T) -> Volatile<'_, T, ReadOnly> {
```
I don't think this and `from_mut_ref` are sound, but I am not quite sure.

phil-opp replied:
Hmm, I'm not sure about this either. I thought the &T/&mut T references are no longer live while the `Volatile` is active so that we don't break any aliasing guarantees. But as I said, I'm not sure about this...

phil-opp replied:
(no longer applies to the latest version)

bjorn3 reviewed (commented):
Apart from `from_ref` and `from_mut_ref` I think the new interface doesn't invoke any UB.

bjorn3 commented:
@RalfJung volatile accesses to `&UnsafeCell<T>` are fine, right?

RalfJung commented:
> volatile accesses to &UnsafeCell<T> are fine, right?

I mean, the access itself is fine, but if you are doing MMIO then you don't want to use a reference. References are considered "dereferencable" and the compiler may insert spurious reads. See e.g. [here](https://github.com/rust-lang/unsafe-code-guidelines/issues/265#issuecomment-761587844) for more details.

phil-opp committed:
Use `*mut T` instead of `&UnsafeCell<T>` since the latter is dereferencable

phil-opp commented:
@RalfJung Thanks for taking a look! I pushed a new commit that switches from `&UnsafeCell<T>` to `*mut T` as the base type. Thus it should no longer be considered as dereferencable, right?



> See e.g. [here](https://github.com/rust-lang/unsafe-code-guidelines/issues/265#issuecomment-761587844) for more details.

The problem with pushing the volatile to the lowest level (e.g. struct fields) is that it doesn't work well for slices/arrays. For example, we don't want to write each byte of a `[u8]` slice through a separate volatile instruction since this would harm performance.

What I'm trying with this crate is to allow putting the volatile wrapper around on the outer level and then providing methods to narrow it down to specific slice ranges (through `Index`/`IndexMut` impls) or struct fields (through the `addr_of_mut!` macro). I'm also trying to provide volatile versions of slice fill/copy methods using intrinsics such as `volatile_copy_nonoverlapping_memory` and `volatile_set_memory`.

RalfJung commented:
> I pushed a new commit that switches from &UnsafeCell<T> to *mut T as the base type. Thus it should no longer be considered as dereferencable, right?

Yes. Basically, "use raw pointers and only raw pointers" is currently the only option for spec-compliant MMIO in Rust.

But I'm really not the expert here, it'd be great if @Lokathor could take a look. :)

Lokathor commented:
I will have a look later today if I can get the chance, but if you want a comparison point you can look at yourself the `voladdress` crate is currently what I use on the GBA. It provides abstraction types for single addresses, compact block of addresses, and strided series of addresses.

Edit: oh, you've probably already read my crate though XD

Lokathor commented on src/lib.rs:
```diff
@@ -192,8 +125,8 @@ where
     {
-        // UNSAFE: Safe, as we know that our internal value exists.
-        unsafe { ptr::read_volatile(&*self.reference) }
+        // UNSAFE: Safe, as ... TODO
```
The `voladdress` crate actually manages to use some type bound shenanigans to make the wrapping method safe or not, see this part: https://github.com/rust-console/voladdress/blob/main/src/plain_voladdress.rs#L141-L190

This way, the safety requirements fall to the declaration site, not on the read and write methods themselves.

phil-opp replied:
That's an interesting approach! I like that you incorporate the access permissions in it directly (i.e. using `()` for "no access" ). The drawback is that it makes the type signatures more complex.

For this crate, it might work to integrate these properties into the `access` parameter. I'll look into it, thanks for the tip!

phil-opp replied:
I pushed a prototype for `unsafe` read/write support using a new `access::Custom` type in https://github.com/rust-osdev/volatile/pull/22/commits/682dd7092e8fc0cc7226db5cfaf4208d09530321.

phil-opp committed:
Add access types for allowing only unsafe reads and/or writes

josephlr commented on src/lib.rs:
```diff
@@ -9,24 +9,22 @@
 #![cfg_attr(feature = "unstable", feature(core_intrinsics))]
 #![cfg_attr(feature = "unstable", feature(const_generics))]
```
Is this still needed now that const generics w/ `usize` is stable?

phil-opp replied:
It isn't, fixed in 0699e4e, thanks!

josephlr commented on src/lib.rs:
````diff
@@ -78,95 +81,27 @@ impl<R> Volatile<R> {
     /// assert_eq!(volatile.read(), 1);
     /// ```
-    pub const fn new(reference: R) -> Volatile<R> {
+    pub unsafe fn new<A>(pointer: *mut T, access: A) -> Volatile<T, A> {
````
Note taking access by value here means that you can't make this method const, is that desirable?

Lokathor replied:
Yeah, you should probably go with a design that allows `const fn`. Details of how are not that important, but `const fn` really feels like a must.

phil-opp replied:
Agreed, I think I found a good solution for this.

josephlr commented:
It seems like you could simplify the marker traits/structs to avoid duplication (and making things like `ReadWrite` just be an alias for a more generic type):
```rust
pub struct NoAccess;
pub struct UnsafeAccess;
pub struct SafeAccess;

pub trait Unsafe {}
pub trait Safe: Unsafe {}

impl Unsafe for UnsafeAccess {}
impl Unsafe for SafeAccess {}
impl Safe for SafeAccess {}

pub struct Access<R, W> {
    pub read: R,
    pub write: W,
}

pub type ReadOnly = Access<SafeAccess, NoAccess>;
pub type WriteOnly = Access<NoAccess, SafeAccess>;
pub type ReadWrite = Access<SafeAccess, SafeAccess>;
```

Then the implementation would be fairly straightforward:
```rust
impl<T: ?Sized> Volatile<T> {
    pub const unsafe fn new<A>(pointer: *mut T) -> Volatile<T, A> {
        Volatile { pointer, access: PhantomData }
    }
    pub const unsafe fn from_ptr(pointer: *const T) -> Volatile<T, ReadOnly> {
        Volatile::new(pointer as *mut T)
    }
    pub const unsafe fn from_mut_ptr(pointer: *mut T) -> Volatile<T, ReadWrite> {
        Volatile::new(pointer)
    }
}

impl<T, Read, Write> Volatile<T, Access<Read, Write>> {
    pub unsafe fn read_unsafe(&self) -> T
    where
        Read: Unsafe,
    {
        ptr::read_volatile(self.pointer)
    }

    pub fn read(&self) -> T
    where
        Read: Safe,
    {
        unsafe { self.read_unsafe() }
    }

    pub unsafe fn write_unsafe(&self, val: T)
    where
        Write: Unsafe,
    {
        ptr::write_volatile(self.pointer, val)
    }

    pub fn write(&self, val: T)
    where
        Write: Safe,
    {
        unsafe { self.write_unsafe(val) }
    }

    pub unsafe fn update_unsafe(&mut self, f: impl FnOnce(&mut T))
    where
        Read: Unsafe,
        Write: Unsafe,
    {
        let mut value = self.read_unsafe();
        f(&mut value);
        self.write_unsafe(value);
    }

    pub fn update(&mut self, f: impl FnOnce(&mut T))
    where
        Read: Safe,
        Write: Safe,
    {
        unsafe { self.update_unsafe(f) }
    }
}
```

Lokathor commented on src/lib.rs:
````diff
@@ -427,55 +342,18 @@ where
     /// assert_eq!(subslice.index(0).read(), 2);
     /// ```
-    pub fn index<'a, I>(&'a self, index: I) -> Volatile<&'a I::Output, A>
+    pub fn index<I>(&self, index: I) -> Volatile<I::Output, ReadOnly>
````
big shame that this isn't `const fn`.

phil-opp replied:
It is! Unfortunately, there's not much that we can do about that since we're using the non-const `get_unchecked_mut` function (see https://github.com/rust-lang/rust/issues/74265).

Lokathor replied:
I guess I don't entirely see why you're using [T] at all instead of just making your own types, but I guess that's up to you.

phil-opp replied:
The problem isn't really the `[T]`, but that we use the [`SliceIndex`](https://doc.rust-lang.org/stable/core/slice/trait.SliceIndex.html) trait for `index`. Unfortunately, trait impls cannot be const yet, but there is an open RFC for this at https://github.com/rust-lang/rfcs/pull/2632. The nice thing about using `SliceIndex` is that we can support all the range operators too.

For only `usize` indexing, we can add a manual const implementation. I just created such a method in https://github.com/rust-osdev/volatile/pull/22/commits/b525f383af7db00f392bd78e5b05390c252acce9.  It requires a few more feature gate, but it should work.

Lokathor replied:
That all still looks way more complicated than I was expecting. I don't get why `map_const` is being called.

When I did it in `voladdress` the block index coded ended up being:

```rust
  /// Indexes to the `i`th position of the memory block.
  ///
  /// ## Panics
  /// * If the index is out of bounds this will panic.
  #[inline]
  #[must_use]
  #[track_caller]
  pub const fn index(self, i: usize) -> VolAddress<T, R, W> {
    if i < C {
      unsafe { self.base.add(i) }
    } else {
      // Note(Lokathor): We force a const panic by indexing out of bounds.
      #[allow(unconditional_panic)]
      unsafe {
        VolAddress::new([usize::MAX][1])
      }
    }
  }
```
And then the same for a Series as well.

I also have this (as non-const because no const traits yet)

```rust
pub fn iter_range<RB: core::ops::RangeBounds<usize>>(
    self, r: RB,
  ) -> VolBlockIter<T, R, W> {
```

but basically I guess I don't currently support sub-slicing or splitting or things like that. The Block and Series types aren't dynamically sized at all, just const generic. If I did want to add dynamically sized regions, my impression is that just having a separate type for it, something like `VolSlice<T: Copy, R, W> { base: VolAddress<_>, length: usize }`, and then you can kinda forget about SliceIndex, which is a little poor to begin with because all the methods are nightly.

Lokathor commented on src/lib.rs:
```diff
@@ -44,10 +42,12 @@ pub mod access;
+where
+    T: ?Sized,
+{
+    pointer: *mut T,
```
might as well go NonNull here

phil-opp replied:
Good idea, thanks!

phil-opp committed:
Remove stabilized const_generics feature

phil-opp commented:
@josephlr 

I thought about removing the `ReadWrite` etc types too, but unfortunately, type aliases cannot be used as type constructors. But methods like the following could work:

```rust
pub fn read_only() -> Access<SafeAccess, NoAccess> {
    Access {
        read: SafeAccess,
        write: NoAccess,
    }
}
```

phil-opp committed:
Redesign access types and adjust (doc)tests

phil-opp committed:
Use `NonNull` instead of `*mut T`

phil-opp committed:
Add lifetime parameter and rename to `VolatilePtr`

The lifetime parameter is required to make `map_mut` sound (otherwise we might get mutable aliasing).

Also:
- add new `from_ref` and `from_mut_ref` contructor methods again.
- add a `new_generic` constructor method to construct a `VolatilePtr` with arbitrary lifetime and access parameters.

phil-opp committed:
Fix: do bounds checking in `index` and `index_mut`

phil-opp committed:
Add `const` index functions under a new `very_unstable` feature

Lokathor commented:
> unfortunately, type aliases cannot be used as type constructors

I'm not sure what this means. Do you mean if you use an alias like

```rust
pub type MyPhantom = PhantomData<u8>;
```
then you can't use `MyPhantom` to make the value? That's easy enough to fix, just add a const.
```rust
#[allow(non_upper_case_globals)]
pub const MyPhantom: MyPhantom = PhantomData;
```

josephlr commented on src/tests.rs:
```diff
@@ -0,0 +1,258 @@
+    // Todo: Custom: no read + unsafe write
+    // Todo: Custom: no read + no write
+
+    // Todo: is there a way to check that a compile error occurs when trying to use
+    // unavailable methods (e.g. `write` when write permission is `UnsafeAccess`)?
```
I think you can have a failing doctest? That might be the easiest way to do it.

See: https://doc.rust-lang.org/edition-guide/rust-2018/rustdoc/documentation-tests-can-now-compile-fail.html

josephlr commented on src/tests.rs:
```diff
@@ -0,0 +1,258 @@
+    volatile.index_mut(..4);
+}
+
+#[cfg(feature = "unstable")]
```
Are there plans to have tests for the `"very_unstable"` features?" Maybe just doc-tests are enough?

josephlr commented on src/lib.rs:
```diff
@@ -70,103 +139,51 @@ impl<R> Volatile<R> {
-            reference,
-            access: PhantomData,
-        }
+    pub const unsafe fn new_with_access<A>(
```
If we were to just delete this function and only have `VolatilePtr::new_generic`, would we really lose much? It seems like you would end up writing about the same amount of code.
```rust
let access = Access {
    read: UnsafeAccess,
    write: SafeAccess,
};
let mut volatile = unsafe { VolatilePtr::new_with_access(NonNull::from(&mut val), access) };
```
vs
```rust
type Custom = Access<UnsafeAccess, SafeAccess>;
let mut volatile = unsafe { VolatilePtr::new_generic::<Custom>(NonNull::from(&mut val)) };
// Or just define inline
let mut volatile: VolatilePtr<_, Access<UnsafeAccess, SafeAccess>> = unsafe { VolatilePtr::new_generic(NonNull::from(&mut val)) };
```

It seems like having to pass the marker type by value isn't really necessary.

josephlr commented on src/access.rs:
```diff
@@ -1,22 +1,48 @@
-pub struct WriteOnly;
-impl Writable for WriteOnly {}
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct Access<R, W> {
+    pub read: R,
+    pub write: W,
+}
+
+impl Access<SafeAccess, NoAccess> {
+    pub const fn read_only() -> ReadOnly {
+        Access {
+            read: SafeAccess,
+            write: NoAccess,
+        }
+    }
+
+    pub const fn write_only() -> WriteOnly {
+        Access {
+            read: NoAccess,
+            write: SafeAccess,
+        }
+    }
+
+    pub const fn read_write() -> ReadWrite {
+        Access {
+            read: SafeAccess,
+            write: SafeAccess,
+        }
+    }
+}
```
If we got rid of the need to pass `Access` by value and instead just referred to the type, we could delete all these type methods, and also just define `Access` like:
```rust
pub struct Access<R, W>(R, W);
```

josephlr reviewed (commented):
This is really looking good, mostly just questions about `new_generic` vs. `new_with_access` I'm not sure it makes sense to have both.

I put together a commit that shows what the crate would look like without `new_with_access`:
https://github.com/josephlr/volatile/commit/952b7cd518ab27a29ca33049d849ba33a2cf41ce

josephlr commented:
> That's easy enough to fix, just add a const.
> 
> ```rust
> #[allow(non_upper_case_globals)]
> pub const MyPhantom: MyPhantom = PhantomData;
> ```

I think we should either:
  - Remove `new_with_access` (and thus the need to even have `Access` values)
    - Then `Access` can just be `pub struct Access<R, W>(R, W);`
  - Define associated constants for `Access` instead of `const fn` methods.
    ```rust
    impl Access<SafeAccess, NoAccess> {
        pub const READ_ONLY: ReadOnly = Access {
            read: SafeAccess,
            write: NoAccess,
        };
        pub const WRITE_ONLY: WriteOnly = Access {
            read: NoAccess,
            write: SafeAccess,
        };
        pub const READ_WRITE: ReadWrite = Access {
            read: SafeAccess,
            write: SafeAccess,
        };
    }
    ```
  - Define a single `Access::new()` method:
    ```rust
    pub struct Access<R, W>(PhantomData<(R, W)>);

    impl<R, W> Access<R, W> {
        const fn new() -> Self {
            Self(PhantomData)
        }
    }
    ```
    - Then you just can use `ReadOnly::new()` when you need a value.

Freax13 commented:
Is it UB to use two `VolatilePtr`'s with `ReadWrite` access pointing to the same address (assuming single threaded execution)?

Lokathor commented:
Nope.

In fact in some cases one could argue that's "the point".

Freax13 commented:
That also implies that `VolatilePtr<T, ReadWrite>` can never be `Send` regardless of `T`.

Lokathor commented:
It's possible for there to be device specific rules, but generally you *can* send and use the pointer from the other thread. Access is not synchronized on its own, you have to rig that up yourself if you're going to use it in both threads at once.

Having it not be send by default and allowing people to override the default is maybe the safe move there.

josephlr commented:
So would that mean that `VolatilePtr<T>` is `Send` and `Sync` iff `T` is `Sync`? (as VolatilePtr<T> behaves like a mutable reference).

Freax13 commented:
`VolatilePtr` doesn't behave like a mutable reference because it's not UB to have two `VolatilePtr`s pointing to the same address. This also means it can't be `Send`: If it were one could create two `VolatilePtr`s pointing to the same address, send one to another thread and create a race condiditon eg. by writing using both pointers from different threads at the same time. This only applies to `ReadWrite` access though.

I think it would be sound to implement `Sync` if `T` is `Sync`.

Lokathor commented:
Precise wording note: Rust allows *race conditions* (which are not UB), just not *data races* (which are UB).

mkroening commented:
Is there anything I can do to help this move forward? :)

josephlr commented on src/lib.rs:
```diff
@@ -7,31 +7,82 @@
-#![cfg_attr(feature = "unstable", allow(incomplete_features))]
+#![cfg_attr(feature = "unstable", feature(slice_ptr_get))]
+#![cfg_attr(feature = "unstable", feature(slice_ptr_len))]
+#![cfg_attr(feature = "very_unstable", feature(const_slice_ptr_len))]
+#![cfg_attr(feature = "very_unstable", feature(const_panic))]
+#![cfg_attr(feature = "very_unstable", feature(const_fn_trait_bound))]
+#![cfg_attr(feature = "very_unstable", feature(const_fn_fn_ptr_basics))]
+#![cfg_attr(feature = "very_unstable", feature(const_trait_impl))]
+#![cfg_attr(feature = "very_unstable", feature(const_mut_refs))]
+#![cfg_attr(feature = "very_unstable", allow(incomplete_features))]
```
Could we drop the `const_fn` stuff and `"very_unstable"` feature for this review. That stuff might be easier to review if we do it in a followup PR. We would probably want to make certain functions `const fn` if a certain feature is enabled, rather than having additional functions. This is what [`x86_64` does](https://github.com/rust-osdev/x86_64/blob/a143355627240bd58f97c4a881b79e86b3ea07cc/src/lib.rs#L16-L45).

josephlr commented on src/lib.rs:
```diff
@@ -607,48 +608,196 @@ where
+        unsafe { self.split_at_mut_unchecked(mid) }
+    }
+
+    unsafe fn split_at_unchecked(
```
Could we add the various "unchecked" methods in a followup PR? This will allow us to focus on having the "right" API for the checked methods. Once we have the "base" PR merged, adding the unchecked methods shouldn't be too bad.

josephlr commented on src/lib.rs:
```diff
@@ -662,23 +811,17 @@ where
-                value,
-                self.reference.len(),
-            );
+            intrinsics::volatile_set_memory(self.pointer.as_mut_ptr(), value, self.pointer.len());
```
It seems like for this method (and all the other methods that use volatile intrinsic), we should just have a stable, normal loop that just does `read_volatile`/`write_volatile` by default.

We could then add a nightly-only `"slice_optimizations"` feature, that changes the implementation to use the intrinsics. Then, when https://github.com/rust-lang/rfcs/pull/2728 gets stabilized, we can switch to unconditionally using the stabilized methods.

josephlr commented on src/lib.rs:
```diff
@@ -688,10 +831,7 @@ where
-where
-    R: Deref<Target = [T; N]>,
-{
+impl<T, R, W, const N: usize> VolatilePtr<'_, [T; N], Access<R, W>> {
```
Maybe I'm confused, but why does `as_slice` need the `'unstable"` feature?

josephlr reviewed (commented):
Some comments (I don't currently have time for a full review)

josephlr commented:
> Is there anything I can do to help this move forward? :)

It's mostly a question of if we can get this PR small enough so that I, @Lokathor, or someone else can review the API.

We would probably also want to prototype some use-cases with this API to make sure things are ergonomic.

josephlr commented:
> `VolatilePtr` doesn't behave like a mutable reference because it's not UB to have two `VolatilePtr`s pointing to the same address. This also means it can't be `Send`: If it were one could create two `VolatilePtr`s pointing to the same address, send one to another thread and create a race condiditon eg. by writing using both pointers from different threads at the same time. This only applies to `ReadWrite` access though.
> 
> I think it would be sound to implement `Sync` if `T` is `Sync`.

I think in the interest of getting this review though, we should postpone discussions of `Send`/`Sync` for a followup issue/PR. For the initial implementation, we should `NonNull`'s lead. The type should be unconditionally `!Send` and `!Sync`. We can always relax the constraints later.

Lokathor commented:
This PR is *rather old* and since it was opened I've experimentally added a "volatile region" type to `voladdress`: https://github.com/rust-console/voladdress/blob/main/src/volregion.rs

So, that's just how I thought things could work. Maybe it can help as a point of reference. I set it up so that it's never a dynamically sized type, because it's a pain in the butt and adds essentially zero value for my purposes.

> The type should be unconditionally !Send and !Sync. We can always relax the constraints later.

This is generally the right call, because (unlike Copy) if someone "knows what they're doing" and wants to override your decision they can just go ahead and do that.

Freax13 commented on src/lib.rs:
```diff
@@ -607,48 +608,196 @@ where
+        unsafe { self.split_at_unchecked(mid) }
+    }
+
+    pub fn split_at_mut(
+        &mut self,
+        mid: usize,
+    ) -> (
+        VolatilePtr<[T], Access<R, W>>,
+        VolatilePtr<[T], Access<R, W>>,
+    ) {
```
Could this method take `Self` as an owned value? A problem with the current implementation is that the returned `VolatilePtr`s can't outlive the reference to the `VolatilePtr`, but they really only need to not outlive the lifetime on the `VolatilePtr`. 

This leads to problems in practice:
```rust
fn fetch_more_data() -> [u8; 4] {
    todo!("fetch some data")
}

fn read_some_data(mut dest: VolatilePtr<[u8], Access<SafeAccess, SafeAccess>>) {
    while dest.len() != 0 {
        let chunk = fetch_more_data();

        let len = cmp::min(chunk.len(), dest.len());
        let (mut head, tail) = dest.split_at_mut(len); // <-- `dest` does not live long enough
        head[..len].copy_from_slice(&chunk[..len]);
        dest = tail; // <-- cannot assign to `dest` because it is borrowed
    }
}
```
(The same code works just fine with `&mut [u8]`.)

josephlr replied:
I've been trying to prototype an updated implementation for this, and I've also realized that having _all_ the methods take `self` is probably what we want to do.

Conceptually, it makes sense for our `VolatilePtr` have `self` methods as it's a `Copy` type. It also reduces the overall number of `mut` vs not-`mut` method variants.

Freax13 commented on src/lib.rs:
````diff
@@ -7,31 +7,82 @@
+/// assert_eq!(field_2.read(), 255);
+/// ```
+#[macro_export]
+macro_rules! map_field {
+    ($volatile:ident.$place:ident) => {
+        unsafe {
+            $volatile.map(|ptr| {
+                core::ptr::NonNull::new(core::ptr::addr_of_mut!((*ptr.as_ptr()).$place)).unwrap()
+            })
+        }
+    };
+}
````
This allows creating `VolatilePtr`s to unaligned fields within packed structs, but `core::ptr::read_volatile` requires pointers to be properly aligned.

phil-opp replied:
Thanks, should be fixed by #28.

Freax13 committed:
add `VolatilePtr::as_slice_mut`

Freax13 committed:
add `VolatilePtr::iter` & `VolatilePtr::iter_mut`

Freax13 committed:
add `VolatilePtr::is_empty`

Freax13 committed:
add missing access where bounds

Freax13 committed:
reject mapping to unaligned fields

Freax13 committed:
WIP: owned with borrow

Freax13 committed:
fix UAF in doctest

Freax13 commented:
Is there a good way to make splitting mutable borrows work? We could maybe achieve this by implementing destructuring in a macro and getting access to the fields that way though it's not quite the same a splitting borrows.

Lokathor commented:
Short answer: Not with just a `volatile` type.

Fundamentally, rust considers a mutable borrow exclusive, and volatile is used on things that you **never** have exclusive control over.

So if you want to build some sort of access control layer to the API using the borrow checker, you should build it "above" this base layer of abstraction.

Freax13 commented:
> Fundamentally, rust considers a mutable borrow exclusive, and volatile is used on things that you **never** have exclusive control over.

As of right now, we never assert that we have exclusive control of the memory, but we require mutable references to a `VolatilePtr` to allow writing to the memory location. What I want is a bit different: I'd like to be able to get two `VolatilePtr`s to two different fields of a struct pointed to by a `VolatilePtr`.
Or put even more concretly: I want this code to work somehow:
![image](https://user-images.githubusercontent.com/14952658/177054678-28efbd14-f14e-4cb8-9db1-635740648f94.png)

Lokathor commented:
> but we require mutable references to a VolatilePtr to allow writing to the memory location.

I simply would not write that code to begin with. These pointers should be Copy and they should take `self` in methods.

> Or put even more concretly: I want this code to work somehow

I completely agree, that code _should_ work.

Freax13 commented:
> I simply would not write that code to begin with. These pointers should be Copy and they should take `self` in methods.

I explored writing such an api in #28 though @phil-opp correctly pointed out in https://github.com/rust-osdev/volatile/pull/28#issuecomment-1146810733 that that would require `VolatilePtr<T>` to be `!Sync` irregardless of `T` which we should probably avoid if possible.

Lokathor commented:
Well according to LLVM it's up to the platform how sync a volatile access is or not, so.... *[thumbs up sign while doing a nervous smile]* I guess good luck with that if your platform is multi-core and doesn't let them be sync. That sounds terrible to me.

Freax13 commented:
> Well according to LLVM it's up to the platform how sync a volatile access is or not, so.... _[thumbs up sign while doing a nervous smile]_ I guess good luck with that if your platform is multi-core and doesn't let them be sync. That sounds terrible to me.

According to https://doc.rust-lang.org/core/ptr/fn.read_volatile.html#safety data races caused by volatile reads and writes are UB:
> Just like in C, whether an operation is volatile has no bearing whatsoever on questions involving concurrent access from multiple threads. Volatile accesses behave exactly like non-atomic accesses in that regard. __In particular, a race between a read_volatile and any write operation to the same location is undefined behavior.__

But yeah, I totally get what you're saying though, for some applications (especially single threaded ones) having a volatile pointer be `!Send + !Sync` would be completely fine, but for some applications that's not what you want to do. This would likely also be a problem if we wanted to introduce this into the language: We might need to two different volatile pointer types, one that's `Send + Sync` and one that isn't. Not sure if there's a way to unify both of those into one type.

Lokathor commented:
Hmmmmmm, well LLVM has a [rather more flexible definition](https://llvm.org/docs/LangRef.html#memory-model-for-concurrent-operations):

> Given that definition, R_byte is defined as follows:
> 
> * If R is volatile, the result is target-dependent. (Volatile is supposed to give guarantees which can support `sig_atomic_t` in C/C++, and may be used for accesses to addresses that do not behave like normal memory. It does not generally provide cross-thread synchronization.)
> * Otherwise, if there is no write to the same byte that happens before R_byte, R_byte returns `undef` for that byte.
> * Otherwise, if R_byte may see exactly one write, R_byte returns the value written by that write.
> * Otherwise, if R is atomic, and all the writes R_byte may see are atomic, it chooses one of the values written. See the [Atomic Memory Ordering Constraints](https://llvm.org/docs/LangRef.html#ordering) section for additional constraints on how the choice is made.
> * Otherwise R_byte returns `undef`.

I think the stdlib docs are actually just in the wrong here. Particularly because it makes interrupt handlers and signal handlers literally impossible to write if the target doesn't _have_ atomics.

phil-opp committed:
Merge pull request #28 from Freax13/unsafe-cell-but-owned

various improvements for the new design

Freax13 commented:
> I fully agree that making `Volatile` a `Copy` type makes it easier to use. My main concern with this design is whether we can still make the `read`/`write` methods safe without causing undefined behavior. If not, there would not be a big advantage in using `Volatile` instead of using `core::ptr::read_volatile` directly.
>
> (from https://github.com/rust-osdev/volatile/pull/28#issuecomment-1385588597)


With the current implementation the `read` and `write` methods are sound. However, the soundness depends on the `VolatilePtr` being `!Send + !Sync`.

phil-opp commented:
After reading through this discussion again, I think there are two possible designs. We can either implement `Copy` or `Send`, but not both. Implementing `Copy` makes the type easier to work with in single-thread environments. Implementing `Send` makes it possible to store the type in a `static`, behind a `Mutex`.

Both approaches seem valid to me, so I implemented them both in #29, as a base for further discussion. The question is now: Which type do we prefer? Or does it make sense to provide them both? Please let me know what you think in #29, so that we can finally replace the current unsound implementation.
